arroyo/processing/strategies/collect.py

```diff
@@ -143,38 +141,41 @@ def terminate(self) -> None:
     def join(self, timeout: Optional[float] = None) -> None:
         if self.batch is not None:
             self.batch.join(timeout)
+            self.__next_step.submit(Message(Value(self.batch, self.batch.offsets)))
             logger.info("Completed processing %r.", self.batch)
             self.batch = None
 
 
 class ParallelCollectStep(CollectStep[TPayload]):
     """
     ParallelCollectStep is similar to CollectStep except it allows the closing and reset of the
-    batch to happen in a threadpool. What this allows for is the next batch to start getting
-    filled in while the previous batch is still being processed.
+    batch to happen in a threadpool. This is useful if closing the batch involes IO-bound tasks
+    such as flushing to a remote storage system. The next batch can start getting filled while
+    the previous batch is being closed.
 
-    The threadpool will have only 1 worker since we want to perform writes to clickhouse sequentially
-    so that kafka offsets are written in order.
+    The threadpool has 1 worker since we want to ensure batches are processed sequentially
+    and passed to the next step in order.
     """
 
     def __init__(
         self,
         step_factory: Callable[[], ProcessingStep[TPayload]],
-        commit_function: Callable[[Mapping[Partition, Position]], None],
+        next_step: ProcessingStep[Batch[TPayload]],
```

Comment:
what about allowing to add a next_step in alternative to the commit_function so that this change is backward compatible? 

We have some direct usages of a collect step in snuba 
https://github.com/getsentry/snuba/blob/5c8d59f0aaf59d96d92a037ffba2436d2683b030/snuba/consumers/strategy_factory.py#L105-L125

and sentry
https://github.com/getsentry/sentry/blob/f2f6d64b8c3f231dc5d487f5d0c98269cfae6140/src/sentry/sentry_metrics/consumers/last_seen_updater.py#L192

Reply:
I would actually prefer to just do a breaking change here and get us onto the version we want faster (every strategy must take a next step), CommitOffsets is the only exception. The number of changes is manageable and the fix in each case is just a few characters. (just passing `(CollectOffsets(commit_function))` there instead of `commit_function`. Keeps the code simpler in the future.

Reply:
ok, up to you